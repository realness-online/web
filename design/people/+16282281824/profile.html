<figure itemscope itemtype="/person" itemid='/people/+16282281824'>
  <time itemprop="created_at" datetime="2018-05-10T16:17:39.624Z"></time>
  <time itemprop="updated_at" datetime="2018-05-10T16:17:39.624Z"></time>
  <svg>
    <use itemprop="profile_vector" xlink:href="/static/icons.svg#silhouette"></use>
  </svg>
  <figcaption>
    <p><span itemprop="first_name">Scott</span><span itemprop="last_name">Fryxell</span></p>
    <a itemprop="mobile" data-value="+16282281824"></a>
  </figcaption>
</figure>



<figure itemscope itemtype="/person" itemid='/people/+16282281824'>
  <svg><use itemprop="profile_vector" xlink:href="/static/icons.svg#silhouette"></use></svg>
  <figcaption>
    <p><span itemprop="first_name">Scott</span><span itemprop="last_name">Fryxell</span></p>
    <a itemprop="mobile" data-value="+16282281824"></a>
  </figcaption>
</figure>
