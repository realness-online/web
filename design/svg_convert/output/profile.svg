<svg version="1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 128 128"><path d="M37.8 7.9c-3.1 2-6.1 4.4-6.8 5.4-.9 1.5-.9 1.8.2 1.2.7-.5-.4 1-2.6 3.3-2.5 2.7-3.7 4.6-3.2 5.7.3.9.2 1.4-.3 1.1-1.5-1-4.1 8.2-5.6 19.4-2.1 15-3.8 24-4.7 24-.4 0-.8.8-.8 1.7-.1 1.8-2.2 5.3-3.3 5.3-1.3 0-5.7 5.2-5.2 6.1.6.9 0 1.1-3.5 1.3-1.3.1-1.5 3.5-1.8 22.9L-.1 128h35.9c30.7 0 35.9-.2 35.9-1.5 0-.8-.4-1.2-.9-.9-.4.3-.9-.5-.9-1.8-.1-1.3-.2-3-.3-3.8-.1-.8.3-2.1.7-2.7 1.3-1.8.1-4.3-2.2-4.3-2 0-3.1-1.8-3.1-5 0-1.4 1.7-2.4 6.8-4.1 6.7-2.2 7.4-3.9.9-2.1-2.7.8-3.8.6-4.9-.5-.8-.8-2.4-1.5-3.6-1.6-1.9-.1-2.1-.6-1.7-4.1.3-2.2 1-4.3 1.7-4.7 1.6-1 .2-6.7-2.2-8.7-2.2-1.9-2.6-5.7-1.5-13.7.9-6.4-.9-15.7-3.6-18.1-1.1-1-3-1.9-4.2-2.1-1.2-.2-3.9-.8-6-1.4-3.7-1-3.7-1-3.6-6 .1-2.7.6-5.2 1.1-5.5 1.4-.9.9-5.2-.9-6.9-1-1-1.2-1.9-.6-2.3.5-.4 1.3-2 1.7-3.5.4-1.5 1.6-3.1 2.9-3.6 1.2-.5 4.6-2.3 7.5-4 5.9-3.4 6.9-3.6 14.3-2 3.4.8 6.4 2.3 9 4.6 2.2 2 3.9 3.2 3.9 2.7 0-.4 1.2.3 2.6 1.7l2.7 2.4-.7-3c-.4-1.7-.9-3.8-1.3-4.9-.4-1.3-.2-1.6.8-1 2.5 1.5-3.3-5.6-6.4-8-4.6-3.4-7.1-3.9-11.3-2.1-2.2.9-5 1.3-6.6 1-2.6-.6-3.3 0-3 2.6.2.9-5.8 2-6.9 1.3-.5-.3-.9-1-.9-1.5 0-.6.5-.7 1-.4.6.3 1.5.3 2.1 0 1.7-1.1-2.2-3.2-6.7-3.6-3.2-.3-5.1.2-9.6 3zM47.5 54c.3.5-.7 1.3-2.2 1.7l-2.8.7 3 .8c2.2.5 1.7.6-1.5.2-5.3-.6-5.7-.9-4-2.9 1.4-1.7 6.5-2 7.5-.5z"/><path d="M82.6 22c.4.8 1.3 2.2 2 3 .7.8 2.4 3.7 3.9 6.5 1.4 2.7 2.3 3.9 1.9 2.5-1.2-4.5-2.3-6.5-5.4-10-1.8-1.9-2.8-2.8-2.4-2zM74.6 42.7c-1.2 1.2-.6 2.5 1 1.9.8-.3 1.2-.2.9.4-.3.6.1 1 .9 1 .9 0 1.6-.4 1.6-1 0-1.2-3.5-3.1-4.4-2.3zM72.1 50.1c-1.7 1-3.1 2.4-3.1 3.1 0 2.3 1.7 4.1 3 3.3.8-.5 1.1-.4.6.3-.7 1.1 4.6 4.2 7.3 4.2 1.4 0 1.4-.2.1-1-1.2-.8-1.1-1 .6-1 1.1 0 2.6-.6 3.3-1.3.6-.6 2.4-1.2 3.8-1.2s2.7-.6 2.9-1.3c.1-.7 0-.8-.2-.4-.5 1-7.2-1-8.3-2.4-.3-.5-2-1.6-3.8-2.5-2.9-1.6-3.3-1.6-6.2.2zm1.4 1.9c.3.5-.1 1-1 1s-1.3-.5-1-1c.3-.6.8-1 1-1 .2 0 .7.4 1 1zM94.3 55.5c0 2.2.2 3 .4 1.7.2-1.2.2-3 0-4-.3-.9-.5.1-.4 2.3zM100.6 75.7c-.4 3.8-.9 9.5-1.1 12.8-.8 10.3-3.1 23.9-4.7 27-.9 1.9-1.3 4.7-1 7.7.4 4.2.7 4.8 2.7 4.8 2.1 0 2.4-.5 2.3-3.9 0-2.2.3-4.5.7-5.2.5-.7.4-1-.2-.6-1.6 1-2.3-4.2-1.3-8.8 1.6-7.4 3-18.9 3.6-29.8.7-11.8 0-14.6-1-4zM69 83c0 .5.7 1 1.5 1s1.5-.5 1.5-1c0-.6-.7-1-1.5-1s-1.5.4-1.5 1zM90.1 123.5c-2.3 4-2.5 4.5-1.3 4.5 1 0 4.5-7 3.9-7.7-.3-.2-1.4 1.2-2.6 3.2zM123.7 125.7c-1.4 1.3-.6 2.3 1.9 2.3 2.2 0 2.5-.2 1.4-1.5-1.4-1.6-2.3-1.9-3.3-.8z"/></svg>