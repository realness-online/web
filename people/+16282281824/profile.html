<figure id="profile" itemscope="itemscope" itemtype="/person">
  <time itemprop="created_at" datetime="2018-05-10T16:17:39.624Z"></time>
  <time itemprop="updated_at" datetime="2018-05-10T16:17:39.624Z"></time>
  <svg >
    <use itemprop="profile_vector" xlink:href="/static/icons.svg#silhouette"></use>
  </svg>
  <figcaption>
    <p><span itemprop="first_name">Scott</span><span itemprop="last_name">Fryxell</span></p>
    <a itemprop="mobile" data-value="+16282281824"></a>
  </figcaption>
</figure>
