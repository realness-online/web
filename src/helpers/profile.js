import itemid from '@/helpers/itemid'
import * as firebase from 'firebase/app'
import 'firebase/storage'
export default {
  async load (id) {
    const person = await itemid.load(`${id}/index`)
    if (person) return person
    else return { id }
  },
  async directory (id, type) {
    const storage = firebase.storage().ref()
    const path = `/people/${id}/${type}`
    if (navigator.onLine) {
      return storage.child(path).listAll()
    } else return null
  },
  as_query_id (id = '/+') {
    return id.substring(2)
  },
  as_avatar_id (id = 'avatar_') {
    return `avatar_${this.as_query_id(id)}`
  },
  as_avatar_fragment (id = 'avatar_') {
    return `#${this.as_avatar_id(id)}`
  },
  as_fragment (id) {
    return `#${this.as_query_id(id)}`
  },
  as_phone_number (id = '/+1') {
    return id.substring(3)
  },
  from_phone_number (phone_number) {
    return `/+1${phone_number}`
  },
  from_e64 (e64_number) {
    return `/${e64_number}`
  }
}
