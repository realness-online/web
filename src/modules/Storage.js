import Item from '@/modules/Item'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'

class Storage {
  static hydrate(item_as_string) {
    return document.createRange().createContextualFragment(item_as_string)
  }
  constructor(item_type, selector = `[itemtype="/${item_type}"]`,
    filename = `${item_type}.html`) {
    this.item_type = item_type
    this.selector = selector
    this.filename = filename
    this.metadata = {'contentType': 'text/html'}
  }
  from_storage() {
    let storage_string = localStorage.getItem(this.item_type)
    return Storage.hydrate(storage_string)
  }
  as_list() {
    return Item.get_items(this.from_storage())
  }
  as_object() {
    return Item.get_first_item(this.from_storage())
  }
  save() {
    // save the information to local storage and if appropriate to the server
    let items = document.querySelector(this.selector)
    if (!items) { return false }
    items = items.outerHTML
    localStorage.setItem(this.item_type, items)
    if (['person', 'posts'].includes(this.item_type)) {
      this.persist(items)
    }
    return true
  }
  sync() {
    // check if an 'item_type'.html already exist for this user.
    // If so merge it's content whith what's stored locally
    const url = this.get_server_file_url()
    if (['person', 'posts'].includes(this.item_type) && navigator.onLine) {
      firebase.auth().onAuthStateChanged(user => {
        if (user) {
          const doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
          firebase.storage().ref().child(doc_u_path).getDownloadURL().then((url) => {
            fetch(url).then(response => {
              response.text().then(server_text => {
                const server_as_fragment = Storage.hydrate(server_text)
                let server_as_list = Item.get_items(server_as_fragment)
                console.log(server_as_list)
              })
              let items = [...this.from_storage, ...from_server]
              // console.log(items)
            })
          })
        }
      })
    }
  }

  get_server_file_url() {

  }

  persist(doc_u_ment, doc_u_path) {
    firebase.auth().onAuthStateChanged(user => {
      if (user && navigator.onLine) {
        const file = new File([doc_u_ment], this.filename)
        if (!doc_u_path) {
          doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
        }
        firebase.storage().ref()
          .child(doc_u_path)
          .put(file, this.metadata)
          .catch(console.log.bind(console))
      }
    })
  }
}

export default Storage
export const person_storage = new Storage('person')
export const posts_storage = new Storage('posts', '[itemprop=posts]')
export const activity_storage = new Storage('activity', '[itemprop=activity]')
export const relations = new Storage('relations', '[itemprop=relations]')
export const phonebook = new Storage('phonebook')

//  use this for firebase logging
// .then( snapshot => {
//   console.log('Uploaded', snapshot.totalBytes, 'bytes.')
//   console.log(snapshot.metadata)
//   console.log(snapshot.fullPath)
//   console.log('File available at', snapshot.downloadURL)
// })
