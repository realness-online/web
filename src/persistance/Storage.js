/** @typedef {import('@/types').Item_Id} Item_Id */

// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Local from '@/persistance/Local'
import Large from '@/persistance/Large'
import Cloud from '@/persistance/Cloud'
import Paged from '@/persistance/Paged'
import { current_user, upload } from '@/use/serverless'
import { as_type, as_filename } from '@/utils/itemid'
import { get } from 'idb-keyval'
import { prepare_upload_html } from '@/utils/upload_processor'

export default class Storage {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    this.id = itemid
    this.type = as_type(itemid)
    this.metadata = { contentType: 'text/html' }
  }
}

/** @extends {Storage} */
export class Poster extends Large(Cloud(Storage)) {}

/** @extends {Storage} */
export class Me extends Cloud(Local(Storage)) {
  constructor() {
    super(localStorage.me)
  }
}

/** @extends {Storage} */
export class Relation extends Local(Storage) {
  constructor() {
    super(`${localStorage.me}/relations`)
  }
}

/** @extends {Storage} */
export class Statement extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/statements`)
  }
}

/** @extends {Storage} */
export class Event extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/events`)
  }
}

/** @extends {Storage} */
export class Offline extends Cloud(Storage) {
  async save() {
    const outerHTML = await get(this.id)
    if (outerHTML) await super.save({ outerHTML })
  }
}

/** @extends {Storage} */
export class History extends Cloud(Storage) {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    super(itemid)
  }

  async save(items) {
    // on purpose doesn't call super.save
    if (!items) return false
    if (current_user.value && navigator.onLine) {
      const { compressed, metadata } = await prepare_upload_html(items)
      try {
        await upload(as_filename(this.id), compressed, metadata)
      } catch (e) {
        console.error(e)
        return false
      }
      return true
    }
    return false
  }
}
