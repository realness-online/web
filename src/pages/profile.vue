<template lang="html">
  <section id="profile" class="page" v-bind:class="{me}">
    <header>
      <profile-as-figure :person='person' :me="me" :previous="true"></profile-as-figure>
      <logo-as-link></logo-as-link>
    </header>
    <icon v-show="working" name="working"></icon>
    <my-posts v-if="me"></my-posts>
    <posts-list v-else :posts='posts'></posts-list>
  </section>
</template>
<script>
  import '@/modules/timeago'
  import {person_storage} from '@/modules/Storage'
  import profileAsFigure from '@/components/profile/as-figure'
  import logoAsLink from '@/components/logo-as-link'
  import postsList from '@/components/posts/as-list'
  import myPosts from '@/components/posts/my-list'
  import icon from '@/components/icon'
  import load_mobile_item from '@/mixins/load_mobile_item'
  export default {
    mixins: [load_mobile_item],
    components: {
      profileAsFigure,
      myPosts,
      postsList,
      logoAsLink,
      icon
    },
    created() {
      const mobile = this.$route.params.mobile
      if (mobile) {
        this.get_items_from_mobile(mobile, 'person').then(items => {
          this.working = false
          this.person = items[0]
        })
        this.get_items_from_mobile(mobile, 'posts').then(items => {
          this.posts = items
        })
      } else {
        this.me = true
        this.working = false
        this.person = person_storage.as_object()
      }
    },
    data() {
      return {
        me: false,
        working: true,
        person: {},
        posts: {}
      }
    }
  }
</script>
<style lang='stylus'>
  @require '../style/variables'
  section#profile > svg.working
    margin-top: base-line
</style>
