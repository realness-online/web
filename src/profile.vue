<template lang="html">
  <section id="profile">
    <header>
      <h1>Profile</h1>
    </header>
    <details>
      <summary>
        <h2>Activity list</h2>
      </summary>
      <activity-list :activity="activity"></activity-list>
    </details>
  </section>
</template>
<script>
  import Item from '@/modules/Item'
  import {activity_storage} from '@/modules/Storage'
  import activity_list from '@/components/activity-list'
  import '@/modules/timeago'
  export default {
    components: {
      'activity-list': activity_list
    },
    data() {
      return {
        activity: Item.get_items(activity_storage.from_storage())
      }
    }
  }
</script>
<style lang="stylus"></style>
