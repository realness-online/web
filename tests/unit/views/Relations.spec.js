import { shallow } from 'vue-test-utils'
import Relations from '@/views/Relations'
import profile from '@/helpers/profile'
import { relations_storage } from '@/persistance/Storage'
describe('@/views/Relations.vue', () => {
  const person = {
    id: '/+14151234356',
    first_name: 'Scott',
    last_name: 'Fryxell',
    mobile: '4151234356'
  }
  it('Render relationship information', () => {
    const relations_spy = jest.spyOn(relations_storage, 'as_list').mockImplementation(() => [person])
    const spy = jest.spyOn(profile, 'load').mockImplementation(() => Promise.resolve(person))
    const wrapper = shallow(Relations)
    expect(wrapper.vm.relations.length).toBe(1)
    expect(spy).toBeCalled()
    expect(relations_spy).toBeCalled()
    expect(wrapper.element).toMatchSnapshot()
  })
})
