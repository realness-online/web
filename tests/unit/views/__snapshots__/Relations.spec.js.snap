// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/views/Relations.vue Render relationship information 1`] = `
<section
  class="page"
  id="relations"
>
  <header>
    <router-link
      to="/phone-book"
    >
      <!---->
    </router-link>
     
    <!---->
  </header>
   
  <hgroup>
    <h1>
      Relations
    </h1>
  </hgroup>
   
  <!---->
</section>
`;
